From 818e1277fdafdd15a732235ff8613b8b1cfb8b1f Mon Sep 17 00:00:00 2001
From: Rohan Puri <rohan_puri@xyratex.com>
Date: Tue, 29 Oct 2013 11:35:06 +0530
Subject: [PATCH 114/121] Changes summary : - 1. FIXED Deadlock arising due to
 improper locking in m0_addb_summaries_post().

This fix was part of fixing addb monitoring infrastructure UT, that was intermittently
failing. UT now passes everytime, after this fix.
---
 addb/addb_monitor.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/addb/addb_monitor.c b/addb/addb_monitor.c
index c4190b5..c4522ed 100644
--- a/addb/addb_monitor.c
+++ b/addb/addb_monitor.c
@@ -471,7 +471,6 @@ M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh       *reqh,
 	M0_ASSERT(mon_ctx->amc_stats_conn != NULL);
 	mon_list = &mon_ctx->amc_list;
 
-	m0_mutex_lock(&reqh->rh_addb_monitoring_ctx.amc_mutex);
 	while (scanned < SCANNED_LIMIT && sent < SENT_LIMIT) {
 		struct m0_addb_sum_rec *sum;
 		struct m0_stats_sum    *rec;
@@ -479,7 +478,9 @@ M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh       *reqh,
 
 		/* end of list reached, wrap around or the list is empty */
 		if (mon == NULL) {
+			m0_mutex_lock(&mon_ctx->amc_mutex);
 			mon = addb_mon_tlist_head(mon_list);
+			m0_mutex_unlock(&mon_ctx->amc_mutex);
 			if (mon == NULL ||
 			/* if the entire list was scanned and not a single
 			 * dirty record found, stop. */
@@ -488,7 +489,9 @@ M0_INTERNAL int m0_addb_monitor_summaries_post(struct m0_reqh       *reqh,
 		}
 		++scanned;
 		sum = mon->am_ops->amo_sum_rec(mon, reqh);
+		m0_mutex_lock(&mon_ctx->amc_mutex);
 		mon = addb_mon_tlist_next(mon_list, mon);
+		m0_mutex_unlock(&mon_ctx->amc_mutex);
 		if (sum == NULL || !sum->asr_dirty)
 			continue;
 		++stats_nr;
@@ -520,14 +523,12 @@ err1_injected:
 		       sum->asr_rec.ss_data.au64s_data, SUM_SIZE(sum));
 		sum->asr_dirty = false;
 		m0_mutex_unlock(&sum->asr_mutex);
-		m0_mutex_unlock(&reqh->rh_addb_monitoring_ctx.amc_mutex);
 		/**
 		 * Post summary record on global machine too.
 		 */
 		M0_ADDB_MONITOR_STATS_POST(&m0_addb_gmc,
 					   m0_addb_rec_type_lookup(rec->ss_id),
 					   cv, rec);
-		m0_mutex_lock(&reqh->rh_addb_monitoring_ctx.amc_mutex);
 		if (used == stats_batch) {
 			fop_data->suf_stats.sf_nr = stats_batch;
 			result =
@@ -550,8 +551,6 @@ err1_injected:
 	}
 	fom->pf_mon = mon;
 out:
-	m0_mutex_unlock(&reqh->rh_addb_monitoring_ctx.amc_mutex);
-
 	return result;
 }
 /*
-- 
1.8.3.2

